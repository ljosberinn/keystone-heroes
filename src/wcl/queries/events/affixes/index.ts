import { Affixes } from "@prisma/client";

import type { EventParams } from "..";
import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyBuffStackEvent,
  ApplyDebuffEvent,
  ApplyDebuffStackEvent,
  BeginCastEvent,
  CastEvent,
  DamageEvent,
  HealEvent,
  InterruptEvent,
  RemoveBuffEvent,
  RemoveDebuffEvent,
} from "../types";
import {
  filterExpression as bolsteringFilterExpression,
  getBolsteringEvents,
} from "./bolstering";
import {
  filterExpression as burstingFilterExpression,
  getBurstingEvents,
} from "./bursting";
import {
  filterExpression as explosiveFilterExpression,
  getExplosiveEvents,
} from "./explosive";
import {
  filterExpression as grievousFilterExpression,
  getGrievousEvents,
} from "./grievous";
import {
  filterExpression as necroticFilterExpression,
  getNecroticEvents,
} from "./necrotic";
import {
  filterExpression as quakingFilterExpression,
  getQuakingEvents,
} from "./quaking";
import {
  filterExpression as sanguineFilterExpression,
  getSanguineEvents,
} from "./sanguine";
import {
  filterExpression as spitefulFilterExpression,
  getSpitefulEvents,
} from "./spiteful";
import {
  filterExpression as stormingFilterExpression,
  getStormingEvents,
} from "./storming";
import {
  filterExpression as tormentedFilterExpression,
  getTormentedEvents,
} from "./tormented";
import {
  filterExpression as volcanicFilterExpression,
  getVolcanicEvents,
} from "./volcanic";

type AffixWithEvents = Exclude<
  Affixes,
  | "Awakened"
  | "Beguiling"
  | "Tyrannical"
  | "Fortified"
  | "Prideful"
  | "Raging"
  | "Infested"
  | "Inspiring"
  | "Reaping"
  | "Skittish"
  | "Infernal"
>;

const affixExpressionMap: Record<AffixWithEvents, string[]> = {
  [Affixes.Sanguine]: sanguineFilterExpression,
  [Affixes.Explosive]: explosiveFilterExpression,
  [Affixes.Grievous]: grievousFilterExpression,
  [Affixes.Necrotic]: necroticFilterExpression,
  [Affixes.Volcanic]: volcanicFilterExpression,
  [Affixes.Bursting]: burstingFilterExpression,
  [Affixes.Spiteful]: spitefulFilterExpression,
  [Affixes.Quaking]: quakingFilterExpression,
  [Affixes.Storming]: stormingFilterExpression,
  [Affixes.Bolstering]: bolsteringFilterExpression,
  [Affixes.Tormented]: tormentedFilterExpression,
  [Affixes.Encrypted]: [],
};

export const getAffixExpression = (affixes: Affixes[]): string[] => {
  return affixes
    .filter((affix): affix is AffixWithEvents => affix in affixExpressionMap)
    .flatMap((affix) => affixExpressionMap[affix]);
};

export const filterAffixEvents = (
  allEvents: AllTrackedEventTypes[],
  affixes: EventParams["affixes"]
): {
  explosiveTargetID: number | null;
  events: (
    | ApplyBuffEvent
    | ApplyBuffStackEvent
    | DamageEvent
    | HealEvent
    | InterruptEvent
    | ApplyDebuffStackEvent
    | BeginCastEvent
    | CastEvent
    | ApplyDebuffEvent
    | RemoveBuffEvent
    | RemoveDebuffEvent
  )[];
} => {
  const affixSet = new Set(affixes);

  const { explosiveTargetID, events: explosiveEvents } = getExplosiveEvents(
    allEvents,
    affixSet
  );

  return {
    explosiveTargetID,
    events: [
      // seasonal
      ...getTormentedEvents(allEvents, affixSet),
      // common affixes
      ...getStormingEvents(allEvents, affixSet),
      ...getSpitefulEvents(allEvents, affixSet),
      ...getSanguineEvents(allEvents, affixSet),
      ...getVolcanicEvents(allEvents, affixSet),
      ...getQuakingEvents(allEvents, affixSet),
      ...getBolsteringEvents(allEvents, affixSet),
      ...getBurstingEvents(allEvents, affixSet),
      ...explosiveEvents,
      ...getGrievousEvents(allEvents, affixSet),
      ...getNecroticEvents(allEvents, affixSet),
    ],
  };
};
