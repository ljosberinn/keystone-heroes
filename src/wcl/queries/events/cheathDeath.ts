import type {
  AllTrackedEventTypes,
  ApplyBuffEvent,
  ApplyDebuffEvent,
} from "./types";

export const CHEAT_DEATHS = {
  PODTENDER: {
    name: "Podtender",
    type: ["applybuff", "removebuff"],
    id: 320_224,
    cd: 360,
    icon: "inv_farm_herbseed",
  },
  CHEATING_DEATH: {
    name: "Cheating Death",
    id: 45_182,
    cd: 600,
    icon: "ability_rogue_cheatdeath",
    type: ["applybuff", "removebuff"],
  },
  CHEAT_DEATH: {
    name: "Cheated Death",
    id: 45_181,
    cd: 600,
    icon: "ability_rogue_cheatdeath",
    type: ["applydebuff", "removedebuff"],
  },
  DEPLETED_SHELL: {
    name: "DepletedShell",
    type: ["applydebuff", "removedebuff"],
    cd: 600,
    id: 320_227,
    icon: "inv_farm_herbseed",
  },
  // cauterize
  UNCONTAINED_FEL: {
    name: "Uncontained Fel",
    type: ["applydebuff", "removedebuff"],
    cd: 480,
    id: 209_261,
    icon: "inv_glaive_1h_artifactaldorchi_d_06",
  },
};

export const cheatDeathFilterExpression = `type in ("applydebuff", "applybuff", "removebuff", "removedebuff") and ability.id in (${Object.values(
  CHEAT_DEATHS
)
  .map((ability) => ability.id)
  .join(", ")})`;

const isCheatDeathEvent = (() => {
  const abilities = Object.values(CHEAT_DEATHS);

  return (
    event: AllTrackedEventTypes
  ): event is ApplyDebuffEvent | ApplyBuffEvent => {
    return abilities.some(
      (ability) =>
        ability.type.includes(event.type) && ability.id === event.abilityGameID
    );
  };
})();

export const filterCheatDeathEvents = (
  allEvents: AllTrackedEventTypes[]
): (ApplyDebuffEvent | ApplyBuffEvent)[] => {
  return allEvents.filter(isCheatDeathEvent);
};
