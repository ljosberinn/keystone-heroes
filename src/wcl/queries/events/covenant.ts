import type { CastEvent, AllTrackedEventTypes } from "./types";
import { createIsSpecificEvent } from "./utils";

export const SHARED_COVENANT_ABILITIES = {
  FLESHCRAFT: 324_631,
  SOULSHAPE: 310_143,
  SUMMON_STEWARD: 324_739,
  DOOR_OF_SHADOWS: 300_728,
} as const;

export const generalCovenantExpression = `type in ("cast", "applydebuff") and ability.id in (${Object.values(
  SHARED_COVENANT_ABILITIES
).join(", ")})`;

const isSharedCovenantAbility = createIsSpecificEvent<CastEvent>({
  abilityGameID: Object.values(SHARED_COVENANT_ABILITIES),
  type: "cast",
});

export const filterCovenantCastEvents = (
  allEvents: AllTrackedEventTypes[]
): CastEvent[] => {
  return allEvents.filter(isSharedCovenantAbility);
};
