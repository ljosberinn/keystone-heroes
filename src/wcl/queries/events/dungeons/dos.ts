import type {
  AllTrackedEventTypes,
  ApplyDebuffEvent,
  BeginCastEvent,
} from "../types";
import { createIsSpecificEvent } from "../utils";

export const DOS_URN = 228_626 as const;
export const DOS_URN_OPENING = 340_004 as const;

/**
 * @see https://www.warcraftlogs.com/reports/hFj3wLzrapC4KvZk#fight=2&type=summary&view=events&pins=2%24Off%24%23244F4B%24expression%24type%20%3D%20%22applydebuff%22%20and%20target.type%20%3D%20%22NPC%22%20and%20ability.id%20%3D%20228626
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "hFj3wLzrapC4KvZk") {
 *       fights(fightIDs: [2]) {
 *         startTime
 *         endTime
 *       }
 *        events(startTime: 34007, endTime: 1759795, filterExpression: "type = \"applydebuff\" and target.type = \"npc\" and ability.id = 228626") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applydebuff" and target.type = "npc" and ability.id = ${DOS_URN}`,
  `ability.id = ${DOS_URN_OPENING}`,
];

const isDosUrnEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: DOS_URN,
});

const isDosUrnBeginCastEvent = createIsSpecificEvent<BeginCastEvent>({
  type: "begincast",
  abilityGameID: DOS_URN_OPENING,
});

export const getDOSEvents = (
  allEvents: AllTrackedEventTypes[]
): (ApplyDebuffEvent | BeginCastEvent)[] => {
  return [
    ...allEvents.filter(isDosUrnEvent),
    ...allEvents.filter(isDosUrnBeginCastEvent),
  ];
};
