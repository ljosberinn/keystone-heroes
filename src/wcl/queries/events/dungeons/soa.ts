import type { AllTrackedEventTypes, ApplyDebuffEvent } from "../types";
import { createIsSpecificEvent } from "../utils";

export const SOA_SPEAR = 339_917 as const;

/**
 * @see https://www.warcraftlogs.com/reports/JVnw7NF6hyz8QqX4#fight=20&type=summary&view=events&pins=2%24Off%24%23a04D8A%24expression%24type%20%3D%20%22applydebuff%22%20and%20target.type%20%3D%20%22NPC%22%20and%20ability.id%20%3D%20339917&hostility=1
 * @example
 * ```gql
 * {
 *   reportData {
 *     report(code: "JVnw7NF6hyz8QqX4") {
 *       fights(fightIDs: [20]) {
 *         startTime
 *         endTime
 *       }
 *       events(startTime: 11052027, endTime: 13152414, filterExpression: "type = \"applydebuff\" and ability.id = 339917") {
 *         data
 *       }
 *     }
 *   }
 * }
 * ```
 */
export const filterExpression = [
  `type = "applydebuff" and ability.id = ${SOA_SPEAR}`,
];

const isSoaSpearEvent = createIsSpecificEvent<ApplyDebuffEvent>({
  type: "applydebuff",
  abilityGameID: SOA_SPEAR,
});

export const getSOAEvents = (
  allEvents: AllTrackedEventTypes[]
): ApplyDebuffEvent[] => {
  return allEvents.filter(isSoaSpearEvent);
};
