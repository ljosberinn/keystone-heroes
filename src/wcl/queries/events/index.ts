import { Affixes } from "@prisma/client";

import type { DungeonIDs } from "../../../db/data/dungeons";
import type {
  AllTrackedEventTypes,
  BeginCastEvent,
  DamageEvent,
  DeathEvent,
} from "../events/types";
import { filterAffixEvents, getAffixExpression } from "./affixes";
import { findExplosiveTargetID } from "./affixes/explosive";
import {
  cheatDeathFilterExpression,
  filterCheatDeathEvents,
} from "./cheathDeath";
import {
  generalCovenantExpression,
  filterCovenantCastEvents,
} from "./covenant";
import {
  deathFilterExpression,
  filterEnemyDeathEvents,
  filterPlayerDeathEvents,
} from "./death";
import { getDungeonExpression, filterDungeonEvents } from "./dungeons";
import {
  filterPlayerInterruptEvents,
  interruptFilterExpression,
} from "./interrupts";
import {
  engineeringBattleRezExpression,
  invisibilityFilterExpression,
  filterProfessionEvents,
} from "./professions";
import {
  filterRemarkableSpellEvents,
  remarkableSpellFilterExpression,
} from "./spells";
import { filterTrinkets, trinketsFilterExpression } from "./trinkets";
import { recursiveGetEvents } from "./utils";

export type EventParams = {
  reportID: string;
  startTime: number;
  endTime: number;
  dungeonID: DungeonIDs;
  fightID: number;
  affixes: Affixes[];
};

const generateFilterExpression = ({
  dungeonID,
  affixes,
}: Pick<EventParams, "dungeonID" | "affixes">) => {
  return [
    deathFilterExpression,
    remarkableSpellFilterExpression,
    engineeringBattleRezExpression,
    invisibilityFilterExpression,
    interruptFilterExpression,
    generalCovenantExpression,
    cheatDeathFilterExpression,
    trinketsFilterExpression,
    ...getDungeonExpression(dungeonID),
    ...getAffixExpression(affixes),
  ]
    .flat()
    .map((part) => `(${part})`)
    .join(" or ");
};

export const getEvents = async (
  params: EventParams,
  playerMetaInformation: { actorID: number; class: number }[]
): Promise<{
  allEvents: AllTrackedEventTypes[];
  playerDeathEvents: DeathEvent[];
  enemyDeathEvents: (DeathEvent | BeginCastEvent | DamageEvent)[];
  explosiveTargetID: number | null;
}> => {
  const filterExpression = generateFilterExpression({
    dungeonID: params.dungeonID,
    affixes: params.affixes,
  });

  const allEvents = await recursiveGetEvents<AllTrackedEventTypes>({
    startTime: params.startTime,
    endTime: params.endTime,
    reportID: params.reportID,
    filterExpression,
  });

  const actorIDSet = new Set(
    playerMetaInformation.map((dataset) => dataset.actorID)
  );

  const dungeonEvents = filterDungeonEvents(
    allEvents,
    params.dungeonID,
    actorIDSet
  );
  const { explosiveTargetID, events: affixEvents } = filterAffixEvents(
    allEvents,
    params.affixes
  );
  const professionEvents = filterProfessionEvents(allEvents);
  const remarkableSpellEvents = filterRemarkableSpellEvents(allEvents);
  const playerDeathEvents = filterPlayerDeathEvents(
    allEvents,
    actorIDSet
    // remarkableSpellEvents
  );
  const interruptEvents = filterPlayerInterruptEvents(allEvents);
  const enemyDeathEvents = filterEnemyDeathEvents(
    allEvents,
    actorIDSet,
    params.affixes.includes(Affixes.Explosive)
      ? findExplosiveTargetID(allEvents)
      : null
  );
  const sharedCovenantAbilitiesEvents = filterCovenantCastEvents(allEvents);
  const cheatDeathEvents = filterCheatDeathEvents(allEvents);
  const trinketEvents = filterTrinkets(allEvents);

  return {
    allEvents: [
      ...professionEvents,
      ...dungeonEvents,
      ...affixEvents,
      ...remarkableSpellEvents,
      ...playerDeathEvents,
      ...interruptEvents,
      ...sharedCovenantAbilitiesEvents,
      ...cheatDeathEvents,
      ...trinketEvents,
    ].sort((a, b) => a.timestamp - b.timestamp),
    playerDeathEvents,
    enemyDeathEvents,
    explosiveTargetID,
  };
};
