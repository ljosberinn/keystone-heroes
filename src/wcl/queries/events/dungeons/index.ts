import { DungeonIDs } from "../../../../db/data/dungeons";
import type { AllTrackedEventTypes } from "../../events/types";
import { filterExpression as dosFilterExpression, getDOSEvents } from "./dos";
import { filterExpression as hoaFilterExpression, getHOAEvents } from "./hoa";
import {
  filterExpression as motsFilterExpression,
  getMOTSEvents,
} from "./mots";
import { filterExpression as nwFilterExpression, getNWEvents } from "./nw";
import { filterExpression as pfFilterExpression, getPFEvents } from "./pf";
import { filterExpression as sdFilterExpression, getSDEvents } from "./sd";
import { filterExpression as soaFilterExpression, getSOAEvents } from "./soa";
import { filterExpression as topFilterExpression, getTOPEvents } from "./top";

const dungeonExpressionMap: Record<DungeonIDs, string[]> = {
  [DungeonIDs.DE_OTHER_SIDE]: dosFilterExpression,
  [DungeonIDs.HALLS_OF_ATONEMENT]: hoaFilterExpression,
  [DungeonIDs.PLAGUEFALL]: pfFilterExpression,
  [DungeonIDs.SANGUINE_DEPTHS]: sdFilterExpression,
  [DungeonIDs.SPIRES_OF_ASCENSION]: soaFilterExpression,
  [DungeonIDs.THEATER_OF_PAIN]: topFilterExpression,
  [DungeonIDs.THE_NECROTIC_WAKE]: nwFilterExpression,
  [DungeonIDs.MISTS_OF_TIRNA_SCITHE]: motsFilterExpression,
};

export const getDungeonExpression = (id: DungeonIDs): string[] =>
  dungeonExpressionMap[id];

export const filterDungeonEvents = (
  allEvents: AllTrackedEventTypes[],
  dungeonID: DungeonIDs,
  actorIDSet: Set<number>
): ReturnType<
  | typeof getDOSEvents
  | typeof getHOAEvents
  | typeof getPFEvents
  | typeof getSDEvents
  | typeof getSOAEvents
  | typeof getTOPEvents
  | typeof getNWEvents
  | typeof getMOTSEvents
> => {
  switch (dungeonID) {
    case DungeonIDs.DE_OTHER_SIDE:
      return getDOSEvents(allEvents);
    case DungeonIDs.HALLS_OF_ATONEMENT:
      return getHOAEvents(allEvents);
    case DungeonIDs.PLAGUEFALL:
      return getPFEvents(allEvents, actorIDSet);
    case DungeonIDs.SANGUINE_DEPTHS:
      return getSDEvents(allEvents);
    case DungeonIDs.SPIRES_OF_ASCENSION:
      return getSOAEvents(allEvents);
    case DungeonIDs.THEATER_OF_PAIN:
      return getTOPEvents(allEvents);
    case DungeonIDs.THE_NECROTIC_WAKE:
      return getNWEvents(allEvents);
    case DungeonIDs.MISTS_OF_TIRNA_SCITHE:
      return getMOTSEvents(allEvents);
    default:
      return [];
  }
};
