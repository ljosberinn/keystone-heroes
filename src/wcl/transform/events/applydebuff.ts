import { EventType } from "@prisma/client";

import { BURSTING } from "../../queries/events/affixes/bursting";
import { encryptedDebuffIDs } from "../../queries/events/affixes/encrypted";
import { NECROTIC } from "../../queries/events/affixes/necrotic";
import { tormentedBuffsAndDebuffs } from "../../queries/events/affixes/tormented";
import { CHEAT_DEATHS } from "../../queries/events/cheathDeath";
import { DOS_URN } from "../../queries/events/dungeons/dos";
import { ENVELOPMENT_OF_MISTS } from "../../queries/events/dungeons/mots";
import { SOA_SPEAR } from "../../queries/events/dungeons/soa";
import { TRINKETS } from "../../queries/events/trinkets";
import type { ApplyDebuffEvent } from "../../queries/events/types";
import type { Processor } from "../utils";

const ids = new Set<number>([
  SOA_SPEAR,
  DOS_URN,
  ENVELOPMENT_OF_MISTS,
  BURSTING.debuff,
  NECROTIC,
  ...tormentedBuffsAndDebuffs
    .filter((deBuff) => deBuff.type.includes("applydebuff"))
    .map((debuff) => debuff.id),
  ...Object.values(CHEAT_DEATHS)
    .filter((ability) => ability.type.includes("applydebuff"))
    .map((ability) => ability.id),
  ...Object.values(TRINKETS)
    .filter((ability) => ability.type.includes("applydebuff"))
    .flatMap((ability) => ability.ids),
  ...Object.values(encryptedDebuffIDs),
]);

export const applyDebuffProcessor: Processor<ApplyDebuffEvent> = (
  event,
  { targetPlayerID, targetNPCID, sourcePlayerID }
) => {
  if (ids.has(event.abilityGameID)) {
    return {
      timestamp: event.timestamp,
      eventType: EventType.ApplyDebuff,
      abilityID: event.abilityGameID,
      targetPlayerID,
      targetNPCID,
      sourcePlayerID,
    };
  }

  return null;
};
