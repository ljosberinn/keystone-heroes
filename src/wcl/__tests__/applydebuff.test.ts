import { DOS_URN } from "../queries/events/dungeons/dos";
import { ENVELOPMENT_OF_MISTS } from "../queries/events/dungeons/mots";
import { SOA_SPEAR } from "../queries/events/dungeons/soa";
import type { ApplyDebuffEvent } from "../queries/events/types";
import { applyDebuffProcessor } from "../transform/events/applydebuff";
import allEvents from "./fixtures/allEvents.json";

describe("applydebuff", () => {
  const params = {
    targetPlayerID: null,
    targetNPCID: null,
    sourcePlayerID: null,
    sourceNPCID: null,
  };

  const applyDebuffEvents = allEvents.filter(
    (event): event is ApplyDebuffEvent => event.type === "applydebuff"
  );

  test("tracks SOA Spear", () => {
    const result = applyDebuffEvents
      .filter((event) => event.abilityGameID === SOA_SPEAR)
      .map((event) => applyDebuffProcessor(event, params));

    expect(result).toMatchSnapshot();
  });

  test("tracks DOS Urn", () => {
    const result = applyDebuffEvents
      .filter((event) => event.abilityGameID === DOS_URN)
      .map((event) => applyDebuffProcessor(event, params));

    expect(result).toMatchSnapshot();
  });

  test("tracks MOTS maze failures", () => {
    const result = applyDebuffEvents
      .filter((event) => event.abilityGameID === ENVELOPMENT_OF_MISTS)
      .map((event) => applyDebuffProcessor(event, params));

    expect(result).toMatchSnapshot();
  });

  test("returns null for unknown applydebuff events", () => {
    expect(
      applyDebuffProcessor(
        {
          type: "applydebuff",
          abilityGameID: 1,
          sourceID: 1,
          targetID: 0,
          timestamp: 1,
        },
        params
      )
    ).toBeNull();
  });
});
