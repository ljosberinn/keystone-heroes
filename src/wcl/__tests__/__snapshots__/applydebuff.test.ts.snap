// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applydebuff tracks DOS Urn 1`] = `
[
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 647832,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
  {
    "abilityID": 228626,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 1286270,
  },
]
`;

exports[`applydebuff tracks MOTS maze failures 1`] = `
[
  {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 35495540,
  },
  {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 35495540,
  },
  {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 35495540,
  },
  {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 35495540,
  },
  {
    "abilityID": 323881,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 35495540,
  },
]
`;

exports[`applydebuff tracks SOA Spear 1`] = `
[
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913707,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 11913879,
  },
  {
    "abilityID": 339917,
    "eventType": "ApplyDebuff",
    "sourcePlayerID": null,
    "targetNPCID": null,
    "targetPlayerID": null,
    "timestamp": 12840160,
  },
]
`;
